src/ChannelServer/World/SkillHandler.cs

```diff
@@ -5,129 +5,182 @@
 			foreach (var target in targets)
 			{
-				//Send.ZC_SKILL_MELEE_GROUND(caster, skill, target.Position.X, target.Position.Y, target.Position.Z, null, 0);
 				var monster = (Monster)target;
-				if (monster.TakeDamage(damage, caster, 2))
+				if (monster.TakeDamage(damage, caster, DamageVisibilityModifier.Skill))
 					Send.ZC_SKILL_CAST_CANCEL(caster, target);
 			}
 		}
 	}
 
-	class GroundSkillHandler : IGroundSkillHandler
+	/// <summary>
+	/// Generic Buff Skill Handler
+	/// </summary>
+	public class BuffSkillHandler : IGroundSkillHandler
+	{
+		public void Handle(Skill skill, Character caster, Position castPosition, Position targetPosition)
+		{
+			if (!skill.CanOverheat || skill.IsOverheated)
+			{
+				Send.ZC_COOLDOWN_CHANGED(caster, skill);
+				skill.OverheatCounter = 0;
+			}
+		}
+	}
+
+	/// <summary>
+	/// Generic Ground Skill Handler
+	/// </summary>
+	public class GroundSkillHandler : IGroundSkillHandler
 	{
-		public void Handle(Skill skill, Character caster, Position position)
+		public void Handle(Skill skill, Character caster, Position castPosition, Position targetPosition)
 		{
-			if (skill.Data.OverHeat != 0)
-				skill.OverHeatCount++;
+			var castedRange = castPosition.Get3DDistance(targetPosition);
+			if (castedRange > skill.Data.MaxRange)
+			{
+				Log.Warning("Player {0} casted skill with id {1} farther than max range, casted range: {2} > max range: {3}.", caster.Name, skill.Id, castedRange, skill.Data.MaxRange);
+				return;
+			}
+			// Enabled for testing max range calculations
+			else
+				Log.Info("Player {0} casted skill with id {1} within max range, casted range: {2} > max range: {3}.", caster.Name, skill.Id, castedRange, skill.Data.MaxRange);
+			if (skill.CanOverheat)
+				skill.OverheatCounter++; // Replace with IncreaseOverheat, it's more verbose?
 			Send.ZC_OVERHEAT_CHANGED(caster, skill);
-			if (skill.Data.OverHeat == 0 || skill.Data.OverHeat == skill.OverHeatCount)
+			if (!skill.CanOverheat || skill.IsOverheated)
 			{
 				Send.ZC_COOLDOWN_CHANGED(caster, skill);
-				skill.OverHeatCount = 0;
+				skill.ResetOverheat();
 			}
-			Send.ZC_SKILL_READY(caster, skill, caster.Position, position);
+			Send.ZC_SKILL_READY(caster, skill, caster.Position, targetPosition);
 			switch (skill.Id)
 			{
 				case 20007:
-					caster.Map.GetVisibleMonsters(caster);
-					var targets = caster.Map.GetAttackableMonstersInRange(position.X, position.Y, position.Z, (int)skill.Data.MaxRange).ToList<IEntity>();
+					var targets = caster.Map.GetAttackableMonstersInRange(targetPosition.X, targetPosition.Y, targetPosition.Z, (int)skill.Data.SplashRange);
 					var damage = caster.GetRandomPAtk() + 100;
-					Send.ZC_SKILL_MELEE_GROUND(caster, skill, position.X, position.Y, position.Z, targets, damage);
+					Send.ZC_SKILL_MELEE_GROUND(caster, skill, targetPosition.X, targetPosition.Y, targetPosition.Z, targets, damage);
```

Comment:
Why not simply pass the position to the sender, instead of breaking it up?

Reply:
Regarding the use position vs the broken up float values, should I just default to the position?

Reply:
Unless there's a reason to use separate values, passing the position is usually to be preferred. Less to type, easier to read.